### PR TITLE
SLCORE-1051 Allow null value for branch during Open in IDE requests

### PR DESCRIPTION
**Pull request by nquinquenel**
[SLCORE-1051](https://sonarsource.atlassian.net/browse/SLCORE-1051)

[SLCORE-1051]: https://sonarsource.atlassian.net/browse/SLCORE-1051?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java [hunk 1]
```diff
@@ -103,7 +112,25 @@ public void handle(ClassicHttpRequest request, ClassicHttpResponse response, Htt
...
+              branchToMatch = storedBranches.getMainBranchName();
+            } else {
+              var serverApi = serverApiProvider.getServerApiOrThrow(connectionId);
+              var branches = sonarProjectBranchesSynchronizationService.getProjectBranches(serverApi, showIssueQuery.projectKey, cancelMonitor);
+              branchToMatch = branches.getMainBranchName();
+            }
+          }
+          var localBranchMatchesRequesting = client.matchProjectBranch(new MatchProjectBranchParams(configScopeId, branchToMatch)).join().isBranchMatched();
+          if (!localBranchMatchesRequesting) {
+            client.showMessage(new ShowMessageParams(MessageType.ERROR, "Attempted to show a fix suggestion for a different branch than the one currently checked out." +
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java]: Right, and I don't even know why the message is mentioning "fix suggestion" while it's a simple Open in IDE action for an issue

[the remark this answers is not in the document]
Comment (nquinquenel) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java]: Although I cannot display the currently checked out branch, as we are asking client for verification

File: API_CHANGES.md [hunk 2]
```diff
@@ -4,6 +4,8 @@
 
 * Signature of `org.sonarsource.sonarlint.core.rpc.protocol.backend.file.DidUpdateFileSystemParams#DidUpdateFileSystemParams` was changed
   * Parameter `addedOrChangedFiles` was split into `addedFiles` and `changedFiles`
+* Removed parameter `branch` from `org.sonarsource.sonarlint.core.rpc.protocol.client.issue.IssueDetailsDto` as it should not be used anymore by the client.
+  This is used as part of the Open in IDE feature, and SLCORE is now checking for branches.
```
Comment (sophio-japharidze-sonarsource) [inline API_CHANGES.md:8]: I think it should mention both Open in IDE & Open Fix in IDE?

**Review by sophio-japharidze-sonarsource — COMMENTED** [review]
If I understood correctly, the same changes would be needed for opening fix suggestions.

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java [hunk 3]
```diff
@@ -103,7 +112,25 @@ public void handle(ClassicHttpRequest request, ClassicHttpResponse response, Htt
...
         if (configScopeId != null) {
-          showIssueForScope(connectionId, configScopeId, showIssueQuery.issueKey, showIssueQuery.projectKey, showIssueQuery.branch,
+          var branchToMatch = showIssueQuery.branch;
+          if (branchToMatch == null) {
+            var branchesStorage = storageService.binding(new Binding(connectionId, showIssueQuery.projectKey)).branches();
+            if (branchesStorage.exists()) {
+              var storedBranches = branchesStorage.read();
+              branchToMatch = storedBranches.getMainBranchName();
+            } else {
+              var serverApi = serverApiProvider.getServerApiOrThrow(connectionId);
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java:122]: Just had a discussion with Tobi on that and I'll include the changes as well for fix suggestions. Just checking that it works for issues for now

File: rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/issue/IssueDetailsDto.java [hunk 4]
```diff
@@ -40,13 +39,12 @@ public class IssueDetailsDto {
   private final TextRangeDto textRange;
 
   public IssueDetailsDto(TextRangeDto textRange, String ruleKey, String issueKey,
-    Path ideFilePath, String branch, @Nullable String pullRequest, String message, String creationDate,
+    Path ideFilePath, @Nullable String pullRequest, String message, String creationDate,
```
[the remark this answers is not in the document]
Comment (nquinquenel) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/issue/IssueDetailsDto.java:42]: cc @thahnen , is it needed for Eclipse?

[the remark this answers is not in the document]
Comment (thahnen) [reply rpc-protocol/src/main/java/org/sonarsource/sonarlint/core/rpc/protocol/client/issue/IssueDetailsDto.java:42]: @nquinquenel: Just checked, not at all :smile:

File: backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (thahnen) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java]: Just a question: why are we asking the client for the currently checked-out branch? We have the Git tools on hand on SLCORE already, couldn't this be done there? :question:

[the remark this answers is not in the document]
Comment (nquinquenel) [reply backend/core/src/main/java/org/sonarsource/sonarlint/core/embedded/server/ShowIssueRequestHandler.java]: Not exactly sure, when we did this we did not have git tools AFAIR. I'd prefer not to refactor everything if it works

**Review by sophio-japharidze-sonarsource — APPROVED** [review]
🔥

**Pull request merged**